demoapp/main.go

```diff
@@ -0,0 +1,228 @@
+	config := loadConfig(ctx, logger)
+	startHTTPServer(ctx, logger, config)
+}
+
+func setupLogger() *slog.Logger {
+	handler := slog.NewJSONHandler(os.Stdout, nil)
+	return slog.New(handler)
+}
+
+// setupTelemetry initializes and configures OpenTelemetry tracing and metrics providers.
+// It creates gRPC-based exporters for tracing and metrics, sets up the TracerProvider and MeterProvider,
+// and returns them for further use.
+//
+// Parameters:
+// ctx (context.Context): The context for the setup process.
+//
+// Returns:
+// tracerProvider (*trace.TracerProvider): The configured TracerProvider for tracing.
+// meterProvider (*metric.MeterProvider): The configured MeterProvider for metrics.
+func setupTelemetry(ctx context.Context) (*trace.TracerProvider, *metric.MeterProvider) {
+	// Create a gRPC-based exporter for tracing
+	expTracer, err := otlptracegrpc.New(ctx)
+	if err != nil {
+		slog.Error("Error in setupTelemetry", "error", err)
+		return nil, nil
+	}
+
+	// Set up the TracerProvider with the created exporter
+	tracerProvider := trace.NewTracerProvider(trace.WithBatcher(expTracer))
+	otel.SetTracerProvider(tracerProvider)
+
+	// Create a gRPC-based exporter for metrics
+	expMeter, err := otlpmetricgrpc.New(ctx)
+	if err != nil {
+		panic(err)
+	}
+
+	// Set up the MeterProvider with the created exporter
+	meterProvider := metric.NewMeterProvider(metric.WithReader(metric.NewPeriodicReader(expMeter)))
+	otel.SetMeterProvider(meterProvider)
+
+	return tracerProvider, meterProvider
+}
+
+// shutdownTelemetry is responsible for gracefully shutting down the OpenTelemetry tracing and metrics providers.
+// It takes a context, the TracerProvider, and the MeterProvider as parameters.
+// It shuts down the TracerProvider and the MeterProvider by calling their respective Shutdown methods.
+// If an error occurs during the shutdown process, it panics.
+//
+// Parameters:
+// ctx (context.Context): The context for the shutdown process.
+// tracerProvider (*trace.TracerProvider): The TracerProvider to be shut down.
+// meterProvider (*metric.MeterProvider): The MeterProvider to be shut down.
+func shutdownTelemetry(ctx context.Context, tracerProvider *trace.TracerProvider, meterProvider *metric.MeterProvider) {
+	// Shutdown the TracerProvider
+	if err := tracerProvider.Shutdown(ctx); err != nil {
+		panic(err)
+	}
+
+	// Shutdown the MeterProvider
+	if err := meterProvider.Shutdown(ctx); err != nil {
+		panic(err)
+	}
+}
+
+// loadConfig loads the configuration from the specified directory and returns a DemoConfig object.
+// It also sets up a watcher to monitor for changes in the configuration file and logs any errors encountered.
+//
+// Parameters:
+// ctx (context.Context): The context for the configuration loading process.
+// logger (*slog.Logger): The logger to be used for logging.
+//
+// Returns:
+// *democonfig.DemoConfig: The loaded DemoConfig object.
+func loadConfig(ctx context.Context, logger *slog.Logger) *democonfig.DemoConfig {
```

Comment:
Our new best practice is to use `atomic.Pointer[demoapp.DemoConfig]` as the value returned.
This will prevent race conditions on config updates.

Reply:
@smintz Not sur ethat I understand the needed change